muffnn/autoencoder/autoencoder.py

```diff
@@ -0,0 +1,686 @@
+        cross-entropy.
+
+    Attributes
+    ----------
+    graph_ : tensorflow.python.framework.ops.Graph
+        The TensorFlow graph for the model.
+    input_layer_size_ : int
+        The dimensionality of the input (i.e., number of features).
+
+    Methods
+    -------
+    fit : Fit the autoencoder.
+    partial_fit : Update the fit of the autoencoder on a mini-batch.
+    transform : Encoded input data.
+    inverse_transform : Decode input data.
+    score : Return the mean score for input data.
+    score_samples : Return the score for each element in input data.
+
+    Notes
+    -----
+    `Adam
+    <https://www.tensorflow.org/versions/r0.8/api_docs/python/train.html#AdamOptimizer>`
+    is used for optimization.
+    Xavier initialization (
+    `<https://www.tensorflow.org/versions/r0.8/api_docs/python/contrib.layers.html#xavier_initializer>`
+    ) is used for the weights.
+    """
+
+    def __init__(self, hidden_units=(16,), batch_size=128, n_epochs=5,
+                 dropout=None, hidden_activation=tf.nn.sigmoid,
+                 output_activation=tf.nn.sigmoid, random_state=None,
+                 learning_rate=1e-3, metric='mse', binary_indices=None,
+                 categorical_indices=None):
+        self.hidden_units = hidden_units
+        self.batch_size = batch_size
+        self.n_epochs = n_epochs
+        self.dropout = dropout
+        self.hidden_activation = hidden_activation
+        self.output_activation = output_activation
+        self.random_state = random_state
+        self.learning_rate = learning_rate
+        self.metric = metric
+        self.binary_indices = binary_indices
+        self.categorical_indices = categorical_indices
+
+    def _set_up_graph(self):
+        """Initialize TF objects (needed before fitting or restoring)."""
+
+        # A placeholder to control dropout for training vs. prediction.
+        self._dropout = tf.placeholder(dtype=tf.float32,
+                                       shape=(),
+                                       name="dropout")
+
+        # Input values.
+        self._input_values = tf.placeholder(tf.float32,
+                                            [None, self.input_layer_size_],
+                                            "input_values")
+        t = self._input_values
+
+        self._default_msk = tf.placeholder(tf.float32,
+                                           [None, self.input_layer_size_],
+                                           "default_msk")
+
+        self._binary_msk = tf.placeholder(tf.float32,
+                                          [None, self.input_layer_size_],
+                                          "binary_msk")
+
+        self._categorical_msks = tf.placeholder(
+            tf.float32,
+            [None, None, self.input_layer_size_],
+            "categorical_msks")
+
+        # Fan in layers.
+        for i, layer_sz in enumerate(self.hidden_units):
+            # Don't dropout inputs.
+            if i > 0:
```

Comment:
On second thought, I don't see the issue with only using dropout on the decoding part. The point here is to avoid overfitting and the idea of rescaling inputs (like one-hot encoded stuff), doesn't make sense to me. Using dropout only in the decoding part seems just fine to me.